python/pylibcudf/pylibcudf/column.pyx

```diff
@@ -360,6 +377,110 @@ cdef class Column:
             []
         )
 
+    @classmethod
+    def from_ndarray(cls, obj):
+        """
+        Create a Column from any object which supports the NumPy array interface.
+
+        Parameters
+        ----------
+        obj : Any
+            The input array to be converted into a `pylibcudf.Column`.
+
+        Returns
+        -------
+        Column
+
+        Raises
+        ------
+        TypeError
+            If the input type is neither `numpy.ndarray` nor `cupy.ndarray`.
+        ImportError
+            If NumPy or CuPy is required but not installed.
+
+        Notes
+        -----
+        If `obj` is a 2D CuPy array, the resulting column is a list column.
+        NumPy conversion logic is not yet implemented.
+        Multi-dimensional arrays (ndim > 2) are not supported.
+        """
+        return cls._from_ndarray(obj)
+
+    @singledispatchmethod
+    @classmethod
+    def _from_ndarray(cls, obj):
+        if np_error is not None:
+            raise np_error
+        if cp_error is not None:
+            raise cp_error
+        raise TypeError(f"Cannot convert a {type(obj)} to a pylibcudf Column")
+
+    if np is not None:
+        @classmethod
+        def from_numpy_array(cls, object obj):
+            # TODO: Should expand to support __array_interface__
+            raise NotImplementedError(
+                "Converting to a pylibcudf Column from "
+                "a numpy object is not yet implemented."
+            )
+
+        @_from_ndarray.register(np.ndarray)
+        @classmethod
+        def _(cls, obj):
+            return cls.from_numpy_array(obj)
+
+    if cp is not None:
+        @classmethod
+        def _from_2d_cupy_array(cls, object arr):
+            """Convert a 2D CuPy array to a Column."""
+            flat_data = arr.ravel()
+
+            num_rows, num_cols = arr.shape
+            offsets = cp.arange(0, (num_rows + 1) * num_cols, num_cols, dtype=cp.int32)
+
+            data_view = gpumemoryview(flat_data)
+            offsets_view = gpumemoryview(offsets)
+            typestr = arr.__cuda_array_interface__['typestr'][1:]
+
+            data_col = cls(
+                data_type=_datatype_from_dtype_desc(typestr),
+                size=flat_data.size,
+                data=data_view,
+                mask=None,
+                null_count=0,
+                offset=0,
+                children=[],
+            )
```

Comment:
I think if the data are C-contiguous we can do this without a copy.

In that case, we can also avoid requiring cupy, because we can make the offsets column with `pylibcudf.filling.sequence`.

Reply:
Data's already required to be C-contiguous. I switched to use `plc.filling.sequence` in https://github.com/rapidsai/cudf/pull/18311/commits/3e5dafac1cdecc4c8442d96f253dcc566a684e8f